### PR TITLE
Simplify time caching in DATE: header formatting (#2176)

### PR DESCRIPTION
**Pull request by socketpair**
## Related issue number

Issue #2176

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [x] Documentation reflects the changes
- [x] If you provide code modification, please add yourself to `CONTRIBUTORS.txt`
  * The format is &lt;Name&gt; &lt;Surname&gt;.
  * Please keep alphabetical order, the file is sorted by names. 
- [x] Add a new news fragment into the `changes` folder
  * name it `<issue_id>.<type>` for example (588.bug)
  * if you don't have an `issue_id` change it to the pr id after creating the pr
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: "Fix issue with non-ascii contents in doctest text files."

### CODE REVIEW

**Comment by socketpair** [comment]
Will fix tests and formatting slightly later. Please review my idea.

File: aiohttp/web_response.py [hunk 1]
```diff
@@ -31,7 +31,27 @@ class ContentCoding(enum.Enum):
...
+_cached_formatted_datetime = None
+def get_formatted_datetime():
+    global _cached_current_datetime
+    global _cached_formatted_datetime
+
+    now = int(time.time())
+    if now != _cached_current_datetime:
+        # Weekday and month names for HTTP date/time formatting;
+        # always English!
+        # Tuples are contants stored in codeobject!
```
Comment (socketpair) [inline aiohttp/web_response.py]: ConStants

File: aiohttp/web_response.py [hunk 2]
```diff
@@ -31,7 +31,27 @@ class ContentCoding(enum.Enum):
 ############################################################
 # HTTP Response classes
 ############################################################
-
+_cached_current_datetime = None
+_cached_formatted_datetime = None
+def get_formatted_datetime():
```
Comment (fafhrd91) [inline aiohttp/web_response.py]: I think this should be helper method.

Comment (socketpair) [reply aiohttp/web_response.py]: Do you mean `@staticmethod`?

Comment (socketpair) [reply aiohttp/web_response.py]: Okay, classmethod

Comment (fafhrd91) [reply aiohttp/web_response.py]: just function in helpers.py

**Comment by asvetlov** [comment]
I have no strong opinion for the PR.
On one hand it simplifies the code, on other it adds global variable.
We have no globals yet and I very like it.
Adding tests might require mocking these global vars (which is not terrible bad but I like to avoid it).

**Comment by socketpair** [comment]
Caching is not actually required.

**Comment by socketpair** [comment]
Well, code coverag is reduced now. But my changes are all tested anyway. I don't mind how I can fix that.

**Comment by socketpair** [comment]
Please wait. I'm writing test for correct caching in the new function.

**Comment by socketpair** [comment]
And I have found and fixed significant bug (!) in my implementation.

**Comment by socketpair** [comment]
Well, I consider this can be merged right now.

File: tests/test_helpers.py [hunk 3]
```diff
@@ -551,3 +505,24 @@ def test_dummy_cookie_jar(loop):
...
     dummy_jar.clear()
+
+
+def test_rfc822_formatted_time_caching(mocker):
+    mock_time = mocker.patch("time.time")
+    mock_gmtime = mocker.patch("time.gmtime")
+    mock_gmtime.side_effect = gmtime
+
+    mock_time.return_value = 1234567890.998
+    assert helpers.rfc822_formatted_time() == 'Fri, 13 Feb 2009 23:31:30 GMT'
```
Comment (socketpair) [inline tests/test_helpers.py]: Friday 13 :)

**Comment by asvetlov** [comment]
Ping

**Comment by socketpair** [comment]
@asvetlov @fafhrd91 

It seems this can be merged.

File: aiohttp/web_request.py [hunk 4]
```diff
@@ -402,11 +400,6 @@ def keep_alive(self):
         """Is keepalive enabled by client?"""
         return not self._message.should_close
 
-    @property
-    def time_service(self):
```
Comment (kxepal) [inline aiohttp/web_request.py:406]: Wonder if that would be a breaking change.

Comment (socketpair) [reply aiohttp/web_request.py:406]: I don't understand what you mean.

Comment (fafhrd91) [reply aiohttp/web_request.py:406]: Should be fine. As I remember time_service is not documented

Comment (kxepal) [reply aiohttp/web_request.py:406]: @socketpair Removing public attribute/method/property may break someone's code who depends on it. 

@fafhrd91 Ah, cool then.

**Pull request merged**